src/pkgcheck/checks/codingstyle.py

```diff
@@ -903,3 +903,118 @@ def feed(self, pkg):
                             lineno=lineno - 1, pkg=pkg)
+
+
+class UnquotedVariable(results.AliasResult, results.Warning):
+    """Variable is used unquoted in a context where it should be quoted.
+
+    Variables like D, FILESDIR, etc may not be safe to use unquoted in some
+    contexts.
+    """
+
+    _name = 'UnquotedVariable'
+
+    def __init__(self, variable, lines, **kwargs):
+        super().__init__(**kwargs)
+        self.variable = variable
+        self.lines = tuple(lines)
+
+    @property
+    def desc(self):
+        s = pluralism(self.lines)
+        lines = ', '.join(map(str, self.lines))
+        return f'unquoted variable {self.variable} on line{s}: {lines} '
+
+
+
+class EbuildUnquotedVariable(UnquotedVariable, results.VersionResult):
+    __doc__ = UnquotedVariable.__doc__
+
+
+class EclassUnquotedVariable(UnquotedVariable, results.EclassResult):
+    __doc__ = UnquotedVariable.__doc__
+
+    @property
+    def desc(self):
+        return f'{self.eclass}: {super().desc}'
+
+
+class _UnquotedVariablesCheck(Check):
+    """Scan files for variables that should be quoted like D, FILESDIR, etc."""
+
+    message_commands = (
+        "die", "echo", "eerror", "einfo", "elog", "eqawarn", "ewarn"
+    )
+    var_names = (
+        "D", "DISTDIR", "FILESDIR", "S", "T", "ROOT", "BROOT", "WORKDIR", "ED",
+        "EPREFIX", "EROOT", "SYSROOT", "ESYSROOT", "TMPDIR", "HOME",
+        # variables for multibuild.eclass
+        "BUILD_DIR",
+    )
+
+    node_types_ok = (
+        # Variable is sitting in a string, all good
+        'string',
+        # Variable is part of a shell assignment, and does not need to be
+        # quoted. for example s=${workdir}/${pn} is ok.
+        'variable_assignment',
+        # Variable sits inside a [[ ]] test command and it's OK not to be quoted
+        'test_command',
+        # Variable is being used in a heredoc body, no need to specify quotes.
+        'heredoc_body',
+    )
+
+    def _var_needs_quotes(self, pkg, node):
+        pnode = node.parent
+        while pnode != node:
+            if pnode.type in self.node_types_ok:
+                return False
+            elif pnode.type == 'command':
+                cmd = pkg.node_str(pnode.child_by_field_name('name'))
+                return cmd not in self.message_commands
+            elif pnode.type in 'array':
+                # Variable is sitting unquoted in an array
+                return True
+            pnode = pnode.parent
+
+        # Default: The variable should be quoted
+        return True
+
+    def _feed(self, item):
+        if item.tree.root_node.has_error:
+            # Do not run this check if the parse tree contains errors, as it
+            # might result in false positives. This check appears to be quite
+            # expensive though...
+            return
+        hits = defaultdict(set)
+        for var_node, _ in bash.var_query.captures(item.tree.root_node):
+            var_name = item.node_str(var_node)
```

Comment:
Multiple ideas:
1. To cache this expensive check, maybe you should add to `pkgcheck.bash.ParseTree` a `@functools.cached_property` of `has_parse_error`, so we compute only once per ebuild / eclass. Currently it won't improve performence as this is the only client of this value, but if we need it in the future?
2. How hard it will be to check if the parse error is done inside `SRC_URI` declaration. I found out that tree-sitter-bash has some issues with our usage of `(` and `)` in `SRC_URI` as part of smart expansions we sometimes do (like `${VALUE:+flag? ( abc )}`). Maybe that way we can decrease the amount of bash parsing failures?

Reply:
> Multiple ideas:
> 
>  1. To cache this expensive check, maybe you should add to `pkgcheck.bash.ParseTree` a `@functools.cached_property` of `has_parse_error`, so we compute only once per ebuild / eclass. Currently it won't improve performence as this is the only client of this value, but if we need it in the future?

My assumption was that `has_parse_error` would be cached in the tree, but that doesn't appear to be the case. I guess it would need to be recomputed for any transformation on the CST.
 
>  2. How hard it will be to check if the parse error is done inside `SRC_URI` declaration. I found out that tree-sitter-bash has some issues with our usage of `(` and `)` in `SRC_URI` as part of smart expansions we sometimes do (like `${VALUE:+flag? ( abc )}`). Maybe that way we can decrease the amount of bash parsing failures?

I think it's doable, but what's the end goal? To avoid reporting unquoted variables in SRC_URI, or find parse errors in tree-sitter-bash? Or something else, I'm missing?

Reply:
> > 1. To cache this expensive check, maybe you should add to `pkgcheck.bash.ParseTree` a `@functools.cached_property` of `has_parse_error`, so we compute only once per ebuild / eclass. Currently it won't improve performence as this is the only client of this value, but if we need it in the future?
> 
> My assumption was that `has_parse_error` would be cached in the tree, but that doesn't appear to be the case. I guess it would need to be recomputed for any transformation on the CST.

Thank you

> > 2. How hard it will be to check if the parse error is done inside `SRC_URI` declaration. I found out that tree-sitter-bash has some issues with our usage of `(` and `)` in `SRC_URI` as part of smart expansions we sometimes do (like `${VALUE:+flag? ( abc )}`). Maybe that way we can decrease the amount of bash parsing failures?
> 
> I think it's doable, but what's the end goal? To avoid reporting unquoted variables in SRC_URI, or find parse errors in tree-sitter-bash? Or something else, I'm missing?

From my random parse failure for ebuild checking, there were ebuilds that had parse failure inside the tree of `SRC_URI` and nothing more. Meaning if we skip the error there, out tree is good enough and then we can have less "unable to check" situations. Of course if this becomes complicated, we can just put this improvement as future task outside of scope for this PR. I do want to include this check in next release.

Reply:
> From my random parse failure for ebuild checking, there were ebuilds that had parse failure inside the tree of `SRC_URI` and nothing more. Meaning if we skip the error there, out tree is good enough and then we can have less "unable to check" situations. Of course if this becomes complicated, we can just put this improvement as future task outside of scope for this PR. I do want to include this check in next release.

Oh, I see what you mean. That's not a bad idea actually. Can I save that for another PR?